cache: switch to Rueidis client for Redis

GiedriusS opened the pull request:
Add support for the Rueidis client. It has been instrumental for us in
avoiding an incident in production from too much network load on the
servers. Since we have lots of alerting/recording rules that use
historical data i.e. from Thanos Store, it means that without
client-side caching the same data is retrieved repeatedly.
Rueidis sends a PTTL command to Redis after retrieving data from Redis
to know what the TTL is.

I contemplated sending a design doc for this but the design doc ended up
very slim. It's either something like this or:

* Adding client-side caching support to go-redis (it's probably a huge
undertaking);
* Using the `type` field to differentiate clients.

I think it's ready to go except for two things:

* There's no way to disable local cache in Rueidis for now due to
limitation in the Rueidis library;
* There's no way to disable the default TTL in case PTTL fails - we
probably don't want to cache sensitive things like index data for longer
then we're supposed to.

I'll try to work on these things if this whole idea of having different
clients look good to you.

Signed-off-by: Giedrius Statkevičius <giedrius.statkevicius@vinted.com>

bwplotka commented on docs/components/store.md:
```diff
@@ -308,10 +310,24 @@ While the remaining settings are **optional**:
+* `GOREDIS` (the default one)
+* `RUEIDIS`
+
+The main difference is that [Rueidis](https://github.com/rueian/rueidis) supports [client-side caching](https://redis.io/docs/manual/client-side-caching/) which is really important. If you have high cardinality metrics then it means that it is no longer necessary to go to remote-object storage to get information about potentially thousands or millions of metrics.
```
> If you have high cardinality metrics then it means that it is no longer necessary to go to remote-object storage to get information about potentially thousands or millions of metrics.

I think this is wrong. The client caching is for Redis client NOT object storage. Also how it helps with cardinality in particular? 

Client side caching to me sounds like extra in-mem cache that is checked first before accessing shared cache in REDIS, right?

yeya24 replied:
Yeah I am also a little bit confused. What's the benefits of client side caching over server side caching. Less network calls?

bwplotka reviewed (approved):
Thanks a lot! One sentence to fix in docs, otherwise LGTM!

matej-g commented:
What's the status on this? Will we proceed with single client as we discussed?

GiedriusS commented:
I talked with @hanjm in private and they depend on RESP2 support in the Redis client so unfortunately we cannot easily from go-redis to Rueidis. So, I propose moving forward with this custom `client` support if that sounds good. Our discussion still stands though - we shouldn't add support for custom clients without good rationale. Let me update the PR and fix the docs.

rueian commented:
> I talked with @hanjm in private and they depend on RESP2 support in the Redis client so unfortunately we cannot easily from go-redis to Rueidis. So, I propose moving forward with this custom `client` support if that sounds good. Our discussion still stands though - we shouldn't add support for custom clients without good rationale. Let me update the PR and fix the docs.

Hi @GiedriusS, just let you know that rueidis now also works in RESP2 with client-side cache disabled.

hanjm commented:
> 
> Hi @GiedriusS, just let you know that rueidis now also works in RESP2 with client-side cache disabled.

Cool.

matej-g commented:
@GiedriusS @hanjm does this mean we can go forward with the original suggestion, i.e. replace clients?

GiedriusS commented:
I think so. I'll update this PR soon.

GiedriusS commented:
Testing out the newest Rueidis client before updating this PR (:

GiedriusS commented:
Replaced go-redis with rueidis and simplified everything since rueidis now supports RESP2 as well. Most notably, this includes support for Redis Enterprise. Also, support for disabling client-side caching has been added so I've adjusted the code for that too. I've been testing this in prod for 3 months+ if not more so I think this should be good to go :muscle: huge thank you goes to @rueian for all of the development and support, and a great library :heart:

douglascamata commented on pkg/cacheutil/redis_client.go:
```diff
@@ -117,19 +119,32 @@ type RedisClientConfig struct {
 	return nil
 }
 
 // RedisClient is a wrap of redis.Client.
```
This comment is now outdated.

douglascamata commented on pkg/cacheutil/redis_client.go:
```diff
@@ -161,39 +176,43 @@ func NewRedisClientWithConfig(logger log.Logger, name string, config RedisClient
-	redisClient := redis.NewClient(opts)
-	if reg != nil {
-		reg = prometheus.WrapRegistererWith(prometheus.Labels{"name": name}, reg)
+	var disableCaching = false
```
Why not `disableCaching := false`? I thought we used `var` only to declare variables and use their type's zero-value.

douglascamata replied:
Also, this is client-side caching, right? Could the variable be named `clientSideCacheEnabled` then? It is more concise with what it really is and avoids the "negative thinking" (which causes double negation when `disableCaching = false`.

douglascamata commented on pkg/cacheutil/redis_client.go:
```diff
@@ -268,42 +273,30 @@ func (c *RedisClient) GetMulti(ctx context.Context, keys []string) map[string][]
+		ctx = timeoutCtx
+	}
+
+	// NOTE(GiedriusS): TTL is the default one in case PTTL fails. 8 hours should be good enough IMHO.
```
What is PTTL?

douglascamata commented on pkg/cacheutil/redis_client.go:
```diff
@@ -219,9 +238,8 @@ func NewRedisClientWithConfig(logger log.Logger, name string, config RedisClient
-	if _, err := c.Set(ctx, key, value, ttl).Result(); err != nil {
-		level.Warn(c.logger).Log("msg", "failed to set item into redis", "err", err, "key", key,
-			"value_size", len(value))
+	if err := c.client.Do(ctx, c.client.B().Set().Key(key).Value(rueidis.BinaryString(value)).ExSeconds(int64(ttl.Seconds())).Build()).Error(); err != nil {
```
Why here we use `Set` with `ExSeconds` being added later instead of `Setex`, like in the `SetMulti` function? Would be cool to be consistent.

douglascamata replied:
Also, wouldn't it be wise to use `DoCache` here to populate the client side cache too? I don't see `DoCache` or `DoMultiCache` being used anymore and from what I understand from Rueidis' README those are functions that will also populate client side cache.

douglascamata replied:
I noticed there is `MGetCache` for getting through client side cache, but I fail to see where we write to the client side cache.

rueian replied:
Hi @douglascamata,

The population of client side cache is done by rueidis when you get response from `DoCache` and `DoMultiCache` with supported commands, such as GET and MGET.

SET is not supported by `DoCache` by design, because the client must be tracked by redis server via cacheable requests to receive related invalidation notifications.

douglascamata replied:
@rueian gotcha, thanks for the clarification! 

I thought it would be useful for us to populate the client side cache whenever we SET something, but it if cost is another GET and by consequence of the design, might not be a good idea.

GiedriusS commented:
Thanks for the review @douglascamata, I've updated the PR.

yeya24 commented on pkg/cacheutil/redis_client.go:
```diff
@@ -117,19 +119,31 @@ type RedisClientConfig struct {
 
+	if c.TLSEnabled {
+		if (c.TLSConfig.CertFile != "") != (c.TLSConfig.KeyFile != "") {
+			return errors.New("both client key and certificate must be provided")
```
So it is either they are both specified or both empty?

yeya24 reviewed (dismissed):
Overall this looks good. Looking forward to merge it!

yeya24 commented:
Btw let's fix the changelog.

bwplotka reviewed (approved):
Amazing work! Worth taking it to 0.30 release as well.
